### PR TITLE
8283269: Improve definition and use of jdk.javadoc.internal.doclets.toolkit.Content

### PR DESCRIPTION
**Pull request by pavelrappo**
This PR started in a draft mode as a refactoring effort. After a few commits and chats with Jonathan Gibbons, I decided that further refactoring belongs to follow-up PRs and that this PR could be marked as "Ready for review".

The motivation for the initial refactoring effort was as follows. The word "tree" is heavily overloaded in the JavaDoc codebase, both in comments and code. Here are some of the contexts this word appears in:  

  * Hierarchy Tree (i.e. overview-tree.html and package-tree.html)
  * DocTree (AST nodes)
  * ClassTree, *TreeWriter (data structures and entities related to supertype-subtype relationship)
  * HtmlTree (HTML nodes) and specifically UL/OL elements which are nested lists

Sometimes contexts overlap, making the word "tree" ambiguous. To reduce this ambiguity, the word "tree" should be dropped in favor of a more specific word of phrase where possible.

In the case where the context is `jdk.javadoc.internal.doclets.toolkit.Content`, the programmer is already aware that they are dealing with trees in the sense that `Content` objects support recursive composition. There's no need to have the phrase "content tree" where "content" would do. Moreover, in the context that is exclusively about `Content` objects, the word "content" can be dropped too, since the type information is assumed.

As an example of content overlap, have a look at the source of `jdk.javadoc.internal.doclets.toolkit.builders.MemberSummaryBuilder::buildSummary`. This method used to needlessly mix DocTree with Content tree.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283269](https://bugs.openjdk.java.net/browse/JDK-8283269): Improve definition and use of jdk.javadoc.internal.doclets.toolkit.Content

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 0ea3ec8c573eb98156ca94319eab709abafa1657

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7843/head:pull/7843` \
`$ git checkout pull/7843`

Update a local copy of the PR: \
`$ git checkout pull/7843` \
`$ git pull https://git.openjdk.java.net/jdk pull/7843/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7843`

View PR using the GUI difftool: \
`$ git pr show -t 7843`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7843.diff">https://git.openjdk.java.net/jdk/pull/7843.diff</a>

</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java [hunk 1]
```diff
@@ -205,95 +205,95 @@ public Boolean visitDeclared(DeclaredType t, Void unused) {
...
      *
      * @param member the member to write parameters for.
      * @param includeAnnotations true if annotation information needs to be added.
-     * @return the content tree containing the parameter information
+     * @return the content containing the parameter information
      */
     protected Content getParameters(ExecutableElement member, boolean includeAnnotations) {
-        Content paramTree = new ContentBuilder();
-        paramTree.add("(");
+        Content paramContent = new ContentBuilder();
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java:227]: Continuing the theme of using more specific words than `content`, I've started using the word `result` for those cases where the code is building up an object that will be the result of the method.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java [hunk 2]
```diff
@@ -205,95 +205,95 @@ public Boolean visitDeclared(DeclaredType t, Void unused) {
...
         }
-        paramTree.add(")");
-        return paramTree;
+        paramContent.add(")");
+        return paramContent;
     }
 
     /**
-     * Get a content tree containing the exception information for the executable member.
+     * Get a content containing the exception information for the executable member.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java:280]: Suggest "Returns content containing ..."

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java [hunk 3]
```diff
@@ -91,9 +91,9 @@ public AbstractMemberWriter(SubWriterHolderWriter writer) {
     /**
      * Adds the summary label for the member.
      *
-     * @param memberTree the content tree to which the label will be added
+     * @param content the content to which the label will be added
      */
-    public abstract void addSummaryLabel(Content memberTree);
+    public abstract void addSummaryLabel(Content content);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java:96]: Earlier in this review you used `target` for such an item

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java [hunk 4]
```diff
@@ -156,39 +156,39 @@ protected void addSummaryLink(TypeElement typeElement, Element member, Content t
...
-            Element member, Content linksTree);
+            Element member, Content target);
 
     /**
      * Returns a link for summary (deprecated, preview) pages.
      *
      * @param member the member being linked to
      *
-     * @return a content tree representing the link
+     * @return a content representing the link
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java:179]: you could either just drop the initial  "a" or even the initial phrase and reduce it to `@return the link`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java [hunk 5]
```diff
@@ -257,51 +257,51 @@ private void addModifier(Element member, Content code) {
...
+        writer.addPreviewInfo(member, content);
     }
 
     protected String name(Element member) {
         return utils.getSimpleName(member);
     }
 
     /**
-     * Adds use information to the documentation tree.
+     * Adds use information to the documentation.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java:298]: An alternative would be one of
* Adds use information to the given content.
* Adds use information to the specified content.

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java:298]: Or, 

```
Adds use information for a list of elements to the specified content.
@param the list of elements
```

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java [hunk 6]
```diff
@@ -400,8 +400,8 @@ public Content getSummaryTableTree(TypeElement tElement) {
     }
 
     @Override
-    public Content getMemberTree(Content memberTree) {
-        return writer.getMemberTree(memberTree);
+    public Content getMember(Content memberContent) {
+        return writer.getMember(memberContent);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractMemberWriter.java:404]: As a comment on the underlying API (not your changes), this method is confusing or under-specified or both.   

I think this method name is too short.  In an `AbstractMemberWriter`, the member might reasonably be expected to be the underlying `Element`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractOverviewIndexWriter.java [hunk 7]
```diff
@@ -62,35 +62,35 @@ public AbstractOverviewIndexWriter(HtmlConfiguration configuration,
...
+        addOverviewComment(target);
+        addOverviewTags(target);
     }
 
     /**
      * Adds the overview comment as provided in the file specified by the
      * "-overview" option on the command line.
      *
-     * @param htmltree the documentation tree to which the overview comment will
+     * @param content the documentation to which the overview comment will
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractOverviewIndexWriter.java:77]: If the parameter name was `target`, you could replace `documentation` with `content`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AllPackagesIndexWriter.java [hunk 8]
```diff
@@ -89,11 +89,11 @@ protected void buildAllPackagesFile() throws DocFileIOException {
     }
 
     /**
-     * Add all the packages to the content tree.
+     * Add all the packages to the content.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AllPackagesIndexWriter.java:92]: Should be `adds`

regarding `the content` ... it seems strange to see `the content` used in a way that does not specify _which_ content. That's why I often use `the given content` or `the specified content` in summary sentences. Then, later on, in the context of a `@param` it works to give the shorter form of `the content`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AnnotationTypeMemberWriterImpl.java [hunk 9]
```diff
@@ -92,24 +92,24 @@ public AnnotationTypeMemberWriterImpl(SubWriterHolderWriter writer,
 
     @Override
     public Content getMemberSummaryHeader(TypeElement typeElement,
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AnnotationTypeMemberWriterImpl.java:94]: OK, this is a weird schizophrenic method ... is it `get` or is it `add` ?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassUseWriter.java [hunk 10]
```diff
@@ -230,24 +230,24 @@ protected void generateClassUseFile() throws DocFileIOException {
     /**
      * Add the class use documentation.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassUseWriter.java:231]: I have a pair of classes I use in small utilities:
1. DeclScanner ... find accessible declarations in a `CompilationUnitTree`
2. (subtype of) DocTreeScanner ... analyze doc comments on those accessible declarations.
Whilst I know you are not a big fan of doc comments on internal API, we could still use them to find doc comments whose first sentence does not begin with 3rd person declarative form (the ...s form) of a verb.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassUseWriter.java [hunk 11]
```diff
@@ -230,24 +230,24 @@ protected void generateClassUseFile() throws DocFileIOException {
...
+            addPackageAnnotationList(c);
         }
-        addClassList(content);
-        contentTree.add(content);
+        addClassList(c);
+        content.add(c);
     }
 
     /**
      * Add the packages elements that use the given class.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassUseWriter.java:246]: Pedantic: `Add the package elements`  (note: package not packages)

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java [hunk 12]
```diff
@@ -194,48 +188,48 @@ protected TypeElement getCurrentPageElement() {
...
         if (!options.noComment()) {
             // Print Information about all the tags here
-            addTagsInfo(typeElement, classInfoTree);
+            addTagsInfo(typeElement, classInfo);
         }
     }
 
     /**
-     * Get the class hierarchy tree for the given class.
+     * Get the class inheritance tree for the given class.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: * Gets

"class inheritance" suggests superclass, which is not a tree. Depending on functional semantics, maybe just "class hierarchy", since superclasses do not form a tree and interfaces form an acyclic graph.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Contents.java [hunk 13]
```diff
@@ -343,7 +343,7 @@
      * a given key in the doclet's resources.
      *
      * @param key the key for the desired string
-     * @return a content tree for the string
+     * @return a content for the string
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Contents.java:346]: Rather than `a content for ...`, consider `content containing ...`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/EnumConstantWriterImpl.java [hunk 14]
```diff
@@ -73,24 +73,24 @@ public void addSummary(Content summariesList, Content content) {
...
 
     @Override
-    public Content getEnumConstantsDetailsTreeHeader(TypeElement typeElement,
-            Content memberDetailsTree) {
-        memberDetailsTree.add(MarkerComments.START_OF_ENUM_CONSTANT_DETAILS);
-        Content enumConstantsDetailsTree = new ContentBuilder();
+    public Content getEnumConstantsDetailsHeader(TypeElement typeElement,
+                                                 Content memberDetails) {
+        memberDetails.add(MarkerComments.START_OF_ENUM_CONSTANT_DETAILS);
+        Content enumConstantsDetailsContent = new ContentBuilder();
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/EnumConstantWriterImpl.java:79]: This is one of those cases where a short, more generic name like `result` might work better

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java (hunk 12 above)
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: I picked "inheritance" over "hierarchy" for consistency, because the relevant entities in HTML and CSS referred to "inheritance"; for example:

 * jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:164
 * jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:934

On the other hand, "hierarchy" was also a good choice because it could be used as a standalone noun, thus eliminating the need for the word "tree".

Thoughts?

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: Yes, my criticism was about the use of the word "tree" in this case. While I agree that this is not the primary use you are trying to eliminate (i.e. HtmlTree-like trees) it  nevertheless is misleading to describe the superclasses of a class as forming a tree, since they form just a single branch of any tree.

**Comment by pavelrappo** [comment]
Jon, just to clarify: the most recent commit (601a17c) does not address any of your earlier comments. They are still to be addressed.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java (hunk 12 above)
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: Is there a suitable term we could use to describe this _ancestry_ information? We are specifically interested in a superclass-subclass chain that connects a class in question with java.lang.Object.

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: `class hierarchy` ?
`superclass chain` ?
`superclasses` ?
`inheritance chain` ?
`list of superclasses` ?
`ancestors` ?


'inheritance` isn't a bad word, it just doesn't work as a noun in this context. The problem is using it with `tree`, since the superclass-subclass chain is a degenerate tree that is a list.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractOverviewIndexWriter.java [hunk 15]
```diff
@@ -62,35 +62,35 @@ public AbstractOverviewIndexWriter(HtmlConfiguration configuration,
...
+     * @param target the documentation to which the overview header will be added
      */
-    protected void addOverviewHeader(Content main) {
-        addConfigurationTitle(main);
-        addOverviewComment(main);
-        addOverviewTags(main);
+    protected void addOverviewHeader(Content target) {
+        addConfigurationTitle(target);
+        addOverviewComment(target);
+        addOverviewTags(target);
```
[the remark this answers is not in the document]
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractOverviewIndexWriter.java:70]: I could revert it if you'd like.

While refactoring, I followed a principle that if a method is specific, then its arguments should hint on that specificity. Conversely, if a method could be used in more than a single context, then its arguments should have general names. For example, the below methods should only be passed the HEAD HTML element, so their respective `HtmlTree` arguments are called "head" rather than "target":

  * jdk.javadoc.internal.doclets.formats.html.markup.Head#addStylesheets
  * jdk.javadoc.internal.doclets.formats.html.markup.Head#addStylesheet
  * jdk.javadoc.internal.doclets.formats.html.markup.Head#addScripts

So it's not that I don't agree with you, it's that sometimes I miss stuff. Which is not that weird given that plethora of tiny, incohesive methods that plague writers.

Since we're here, the way builders and writers work together to assemble an HTML tree is strange [^1]. Although I think that it is a result of evolution, we have to fix it. I'm not ready to have this conversation right now, as I think this PR is already doing way too much beyond its original intent.

[^1]: I hope to investigate alternatives in https://bugs.openjdk.java.net/browse/JDK-8283576.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java (hunk 12 above)
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: If we look at the terms available in JLS, what about "All Extended Classes" in jdk.javadoc codebase **and** the generated documentation? (Currently, "Inheritance Tree" is used as a tooltip title of the respective region.) "All Extended Classes" goes nicely with "All Implemented Interfaces". Alternatively we could use Superclasses and  Superinterfaces as per https://docs.oracle.com/javase/specs/jls/se18/html/jls-8.html#jls-8.1.4.

No matter which of those two we choose, it might feel awkward because of the inherent (no pun intended) asymmetry between classes and interfaces: there can be only one direct superclass but multiple direct superinterfaces.

**Comment by pavelrappo** [comment]
Some stats on the effects of this PR.

Before:

    $ egrep -R '(T|t)ree' src/jdk.javadoc/ | wc -l
        4448
    $ egrep -R '(C|c)ontent' src/jdk.javadoc/ | wc -l
        3934

After:

    $ egrep -R '(T|t)ree' src/jdk.javadoc/ | wc -l
        2820
    $ egrep -R '(C|c)ontent' src/jdk.javadoc/ | wc -l
        3836

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java (hunk 12 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:222]: I think the distinction is between "down" from Object (subtypes form a tree) or "up" to Object (superclasses form a list, super interfaces form a graph)

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TreeWriter.java [hunk 16]
```diff
@@ -158,14 +158,12 @@ protected void addPackageTreeLinks(Content content) {
...
-     *
-     * @return a content for the tree header
+     * {@return a new HTML BODY element}
      */
-    protected HtmlTree getTreeHeader() {
+    private HtmlTree getBody() {
         String title = resources.getText("doclet.Window_Class_Hierarchy");
         HtmlTree bodyTree = getBody(getWindowTitle(title));
-        bodyContents.setHeader(getHeader(PageMode.TREE));
+        bodyContents.setHeader(getHeader(PageMode.TREE)); // idempotent
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TreeWriter.java:166]: why is the comment significant?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Script.java [hunk 17]
```diff
@@ -102,11 +102,11 @@ public Script appendStringLiteral(CharSequence text, char quoteChar) {
      * Returns a "live" view of the script as a {@code Content} object.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Script.java:102]: Don't change it now, but the comment abut a "live view" is sort-of unnecessary because in general most HTMLTree objects are mutable.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java [hunk 18]
```diff
@@ -49,38 +49,38 @@
...
      * Get the header for the summary header.
      *
-     * @return a content with the summary header
+     * @return the summary header
      */
     Content getSummariesList();
 
     /**
-     * Get the header for the summary tree.
+     * Wrap the content into summary section.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ModuleSummaryWriter.java:64]: Wraps

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/SerializedFormWriter.java [hunk 19]
```diff
@@ -68,40 +68,34 @@
     Content getPackageSerializedHeader();
 
     /**
-     * Add the serialized tree per package to the serialized summaries.
+     * Add the serialized package to the serialized summaries.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/SerializedFormWriter.java:71]: Adds

**Review by jonathan-gibbons — APPROVED** [review]
OK ...

Well, there's a huge amount of work here, most of which is a genuine improvement, so well done for all that, but I get the sense this work got out of control a bit, given there are many "themes" in this work, each of which would have been a worthy and significant PR in its own right.  As a result, the whole is much harder to review than the sum of its parts.

The primary theme seems to be to remove the word `tree` from method names and comments when it is either in appropriate or somewhat unnecessary.

Amongst the additional themes that are not directly related to the JBS issue:

* use of `var` 
* removing unnecessary comments from overriding methods
* use of `{@return}`
* minor grammar 

Don't get me wrong: these are all good improvements. It's just that doing them all together makes it harder to see the trees in the forest. (sic) 

You are somewhat fortunate in that there is not a lot of major concurrent development in javadoc right now; the merge could have been horrible!

So, to summarize, thanks for taking all this on; I'll approve it, but please file JBS issues for any follow-up cleanup that you consider still needs to be done.
